src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
     log_println(3, "Successfully returned from run_test() routine");
   } else {
     log_println(3,
-                "Child %d returned non-zero (%d) from "
-                "run_test() results some test failed!",
+                "Child %d returned non-zero (%d) from run_test() results some "
+                "test failed!",
                 getpid(), retcode);
-    child_sig(0);
   }
-  SSL_free(ctl.ssl);
-  close(ctl.socket);
+  close_connection(&ctl);
 #if USE_WEB100
   web100_detach(agent);
 #elif USE_WEB10G
   estats_nl_client_destroy(&agent);
 #endif
-  // log_free(); // Don't free the log we use it all the time
-  // log_println()
-  // Also makes valgrind angry
 
   if (cputime && workerThreadId) {
     cputimeworkerLoop = 0;
     pthread_join(workerThreadId, NULL);
   }
-
-  // At this point the tests have been run and we need to clean up
-  // and handle and child processes that might still be lying
-  // around.  If we don't we get zombies. The pkt-pair handling
-  // created 2 childern and we need to get rid of them. To know
-  // what PIDs to look for on 1/19/10 the run_test() routine was
-  // modified to pass around these values. The values are set in
-  // the proper routine in the testoptions.c file.
-  // Then add call to child_sig() routine and pass in these PID's
-  // so we handle each child in sequence
-
-  log_println(6, "remove pkt-pair children c2s=%d, s2c=%d", testopt.child1,
-              testopt.child2);
-  if (testopt.child1 != 0) child_sig(testopt.child1);
-  if (testopt.child2 != 0) child_sig(testopt.child2);
   exit(0);
 }
 
+int max(int a, int b) { return (a > b) ? a : b; }
+
 /**
- * The server's main loop.  This is the function that, once all arguments are
```

Comment:
This function is 120 lines long and even with explanation I find it pretty confusing. Consider the following refactoring:

```
/**
 * @returns NEW_CHILD_WAITING if the wakeup event was the result of a new child waiting to connect
 */
int wait_for_wakeup_event(int listenfd, int signalfd, int wait_forever) {
  // 1379-1401 move to here
}

struct ndtchild *spawn_new_child(int listenfd, SSL_CTX *ssl_context) {
  // 1403-1483 move to here
  // consider further refactoring all the address tinkering stuff (1404-1423) to
  // its own function
}

void NDT_server_main_loop(SSL_CTX *ssl_context, int listenfd, int signalfd) {
  int wakeup_reason;
  struct ndtchild *head_ptr = NULL, *tail_ptr = NULL, *new_child;
  for (;;) {
    // Wait for a new connection, an interruption, or a timeout.
     wakeup_reason = wait_for_wakeup_event(listenfd, signalfd, (head_ptr == NULL));
    if (wakeup_reason == NEW_CHILD_WAITING) {
      new_child = spawn_new_child(listenfd, ssl_context);
      if (new_child) enqueue_new_child(new_child, &head_ptr, &tail_ptr);
    }
    // Perform queue maintenance: send messages to clients and reap the dead.
    queue_maintenance(&head_ptr, &tail_ptr);
  }
}
```